From 6310dee801af9d366c43dc060cd63506d20c497a Mon Sep 17 00:00:00 2001
From: Bo Shen <voice.shen@atmel.com>
Date: Wed, 3 Dec 2014 16:13:49 +0800
Subject: [PATCH] ARM: atmel: gpio: make it safe to enable pull up or pull down

When enable the PIO pull up function, it is safe to disable
the PIO pull down configuration and vice versa.

Signed-off-by: Bo Shen <voice.shen@atmel.com>
---
 drivers/gpio/at91_gpio.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/gpio/at91_gpio.c b/drivers/gpio/at91_gpio.c
index 6517af1..6959d90 100644
--- a/drivers/gpio/at91_gpio.c
+++ b/drivers/gpio/at91_gpio.c
@@ -46,10 +46,17 @@ int at91_set_pio_pullup(unsigned port, unsigned pin, int use_pullup)
 
 	if (at91_port && (pin < 32)) {
 		mask = 1 << pin;
-		if (use_pullup)
+		if (use_pullup) {
+#ifdef CPU_HAS_PIO3
+			writel(1 << pin, &at91_port->ppddr);
+#endif
 			writel(1 << pin, &at91_port->puer);
-		else
+		} else {
 			writel(1 << pin, &at91_port->pudr);
+#ifdef CPU_HAS_PIO3
+			writel(1 << pin, &at91_port->ppder);
+#endif
+		}
 		writel(mask, &at91_port->per);
 	}
 
-- 
1.9.1

